src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -123,9 +125,33 @@ protected override DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData)
                 return ImmutableArray<IDiagnosticSource>.Empty;
             }
 
+            if (nonLocalDocumentDiagnostics)
+                return GetNonLocalDiagnosticSources();
+
             return taskList
                 ? ImmutableArray.Create<IDiagnosticSource>(new TaskListDiagnosticSource(document, globalOptions))
                 : ImmutableArray.Create<IDiagnosticSource>(new DocumentDiagnosticSource(diagnosticKind, document));
+
+            ImmutableArray<IDiagnosticSource> GetNonLocalDiagnosticSources()
+            {
+                Debug.Assert(!taskList);
+
+                // Compiler does not report any non-local diagnostics, so we bail out when those are requested.
+                // Additionally, all analyzer reported non-local diagnostics are classified as semantic diagnostics,
+                // so we bail out when only analyzer syntax diagnostics are requested.
+                if (diagnosticKind is DiagnosticKind.CompilerSyntax or DiagnosticKind.CompilerSemantic or DiagnosticKind.AnalyzerSyntax)
+                    return ImmutableArray<IDiagnosticSource>.Empty;
+
+                // Non-local document diagnostics are reported only when full solution analysis is enabled.
+                if (globalOptions.GetBackgroundAnalysisScope(document.Project.Language) != BackgroundAnalysisScope.FullSolution)
+                    return ImmutableArray<IDiagnosticSource>.Empty;
+
+                // NOTE: We compute diagnostics using WorkspaceDocumentDiagnosticSource to ensure full project is analyzed to get the reported diagnostics.
+                //       Additionally, we only request non-local document diagnostics here as local document diagnostics are reported with DocumentDiagnosticSource above.
+                return ImmutableArray.Create<IDiagnosticSource>(new WorkspaceDocumentDiagnosticSource(document, ShouldIncludeAnalyzer, includeLocalDocumentDiagnostics: false, includeNonLocalDocumentDiagnostics: true));
```

Comment:
would it make more sense to create a new document diagnostic source for non-local diagnostics?  Might make it easier to see whats going on rather than trying to figure out based on the combination of booleans that get passed into the source.